README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# embrs <img src="man/figures/logo.png" align="right" alt="" width="120" />

<!-- badges: start -->
<!-- badges: end -->

> **embrs** is in development

**embrs** was initially written as part of work on a scoping exercise to
provide early estimates of the environmental impacts of the diesel to
electric bus transition published as [Tivey et al
2023](https://doi.org/10.3390/su15021522)

## Project Webpage

**embrs** [projects pages](https://karlropkins.github.io/embrs/)

## Installation

You can install the development version of embrs from
[Github](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("karlropkins/embrs")
```

## Example

**embrs** uses vehicle and route objects to build emission models in the
classic form:

> \[emissions\] \* \[activity\]

For example:

``` r
library(embrs) 
# a EURO VI ICE diesel bus weighing 15925 kg
bus.1 <- bus_ice(name="E6DV", veh.wt=15925, euro.class="vi", eng.fuel="diesel") 
# a 30 km/hr route
route.30 <- route_veh_spd(30)
# multiple them 
mod <- bus.1 * route.30
# plot that
plot(mod)
```

<img src="man/figures/README-example1-1.png" width="100%" />

Vehicles can also be added together to make fleets and these and routes
can be multiplied to produce small-scale emission inventories or compare
emissions on different routes.

For example:

``` r
# EURO VI diesel versus Battery Electric BUS comparison from Tivey et al (2023)

# a EURO VI ICE diesel bus weighing 15925 kg
bus.1 <- bus_ice(name="E6DV", veh.wt=15925, euro.class="VI", eng.fuel="diesel") 
# a battery electric bus weighing 17725 kg (and conventional brakes) 
bus.2 <- bus_bev(name="BEV", veh.wt=17725) 
# like bus.2 but with regenerative brakes operating at 25% efficiency
bus.3 <- bus_bev(name="BEV\nlow", veh.wt=17725, brk.regen = 0.25) 
# like bus.3 but 75% efficiency
bus.4 <- bus_bev(name="BEV\nhi", veh.wt=17725, brk.regen = 0.75)  
# a small fleet
fleet <- bus.1 + bus.2 + bus.3 + bus.4
# some routes 
routes <- route_naei_urban() + route_naei_rural() + route_naei_motorway()
# an inventory
inventory <- fleet * routes
# just plotting PM contributions by vehicle (bus.1 to 4)
plot(inventory, plot.type="by.vehicle", em.type="just.pm")
```

<img src="man/figures/README-example2-1.png" width="100%" />

## Contributing

Contributions are very welcome. Please see both [contribution
guidance](https://karlropkins.github.io/embrs/CONTRIBUTING.html) and
[code of
conduct](https://karlropkins.github.io/embrs/CODE_OF_CONDUCT.html) if
interested in contributing.

## License

[GPL-3](https://karlropkins.github.io/embrs/LICENSE.html)


NEWS.md
# Version 0.0 - Release Notes

* [0.0.8]
    * added coaches [2023-06-15] 
    * tidied error messaging [2023-06-13]
    * added ef_vein_eea_exhaust wrappers for co, ch4 and nmhc [2023-06-12]
    * add plot.type to plot(embrs); NB embrs paper plot no longer 
    default plot output [2023-06-12]
    * added route_veh_spd [2023-06-12]
    * added ef_vein_eea_exhaust_nox; added em.type option to build_inventory 
    and plot; plot defaults to just showing pm emissions rather than all 
    emissions [2023-06-07]   
    * started restructuring code for non-pm emissions and other vehicle 
    types [2023-06-60] 
* [0.0.7]
    * replaced separate beddows, embrs1 and embrs2 vehicle models with one 
    model with extra argument method = "beddows", "embrs1" or "embrs1" 
    [2023-06-05] 
    * replaced ef_eea2019 with ef_vein_eea [2023-06-02]
    * added verbose argument to all ef functions and build_inventory; 
    added method documentation to EF output [2023-05-26]
    * standardised formal arguments in ef functions [2023-05-26]
    * removed ef_copert and ef_emep related code; ef_eea2019 functions 
    now supersede all ef_copert and ef_emep equivalents [2023-05-26]
    * added bus_dice() as diesel bus shortcut [2023-05-23]
    * added ef_eea2019 like ef_copert5 but using vein::ef_eea() [2023-02-07]
    * now imports vein [2023-02-07] 
* [0.0.6]
    * ef_copert5 update (force setting to set-points) [2023-01-24]
    * archive and web update (news, readme, index) [2023-01-19] 
    * added ef_copert5 functions (still needs work; likely to replace 
      ef_emep) [2023-01-19]
* [0.0.5]
    * standardised errors messages (still needs work) [2022-12-30]
    * added draft favicon [2022-12-30]
* [0.0.4]
    * started embrs emission functions as ef_embrs... [2022-12-27]
    * added naei_route2spd lookup table [2022-12-23]
    * started exhaust emission functions as ef_emep... [2022-12-23]
    * added vehicle/fleet object workforce function [2022-12-21]
* [0.0.3]
    * website update, coc links (and minor fixes) [2022-12-14]
    * website update, news update [2022-12-14]
    * generalised ef_beddows... functions and extended to pm10 [2022-12-14] 
    * website update, updated description (web links) [2022-12-13]
    * website update, added coc and contribution [2022-12-13]
* [0.0.2]
    * object tidying (print.embrs output) [2022-12-13]
    * object handling revision [2022-12-13]
    * website skeleton build [2022-12-12]
* [0.0.1] 
    * first draft embrs object docs [2022-12-12]
    * revised embrs object classes (ops handling) [2022-12-11]
    * first draft embrs object classes [2022-12-09]
    * built basic package using grey.area [2022-09-21]
    * added ef_beddows functions (basic methods from beddows paper) [2022-09-21]
    * added ef_embrs functions (basic methods from bus paper) [2022-09-21]

    



R/embrs.local.R
############################################
#' @title embrs common vehicle builds
############################################

#' @name embrs.local
#' @aliases embrs.vehicles embrs_ice embrs_bev
#' @description Common builds for __embrs__ vehicle builds.
#' @param veh.wt (required numeric) weight of vehicle in kg.
#' @param veh.type (required character) type of vehicle, e.g bus.
#' @param eng.fuel (required character) the engine fuel used by the vehicle,
#' currently only diesel.
#' @param euro.class (required character) vehicle EURO classification,
#' e.g. PRE, I, II, III, IV, V or VI.
#' @param n (numeric) number of vehicles, default 1.
#' @param method (character) vehicle model, currently one of: beddows,
#' embrs1 (default) or embrs2.
#' @param name (character) name of the vehicle object.
#' @param ... other arguments, currently passed on to emission factor models,
#' e.g. ef_embrs..., ef_beddows... and ef_vein_eea... In some cases, these
#' may request additional inputs, e.g. when ef_vein_eea has more than one
#' ef model for a given vehicle/engine/fuel/euro.class combination exhaust
#' emission model it may also request the exhaust technology (exh.tech),
#' see ?ef_vein_eea_exhaust.
#' @note Work in progress; these may change...
#' @returns These functions make common vehicle objects for use in __embrs__
#' emissions models. The main models ice_build and bev_build are for internal
#' compression engine and battery electric vehicles, respectively.
#' @references These functions are based on methods developed and reported by:
#'
#' Beddows, D.C. and Harrison, R.M., 2021. PM10 and PM2.5 emission factors for
#' non-exhaust particles from road vehicles: Dependence upon vehicle mass and
#' implications for battery electric vehicles. Atmospheric Environment, 244,
#' p.117886. \url{https://doi.org/10.1016/j.atmosenv.2020.117886}.
#'
#' And, extrapolated to speed (embrs1) and, break/tyre-wear (embrs2)
#' related emission factors in:
#'
#' Tivey, J., Davies, H.C., Levine, J.G., Zietsman, J., Bartington, S.,
#' Ibarra-Espinosa, S. and Ropkins, K, 2023. Meta-Analysis as Early Evidence on
#' the Particulate Emissions Impact of EURO VI on Battery Electric Bus Fleet
#' Transitions. Sustainability 15, 1522. \url{https://doi.org/10.3390/su15021522}


##############################################
#doing this so we can start generalising
##############################################
#currently intending all vehicle objects to
#be just wrappers for these with veh.type set/forced???

#veh types
#done: bus
#doing: coach
#proposing: truck (hgv), van (n1 to n3), car, motorcycle,
#but large job to get right...


#eng type
#done: ice, bev,
#doing:
#proposing: hybrid, bifuel?

##################
#to think about
####################

#gas phase species handling

#moving _beddows(), _embrs1(), _embrs2() into parent function
#e.g. model = beddows
#it would reduce the number of functions/vehicle type

#splatted function
#' @rdname embrs.local
#' @export

embrs_ice <-
  function(veh.wt = NULL, veh.type = NULL, eng.fuel = NULL, euro.class = NULL,
           n = 1, method = "embrs1", name = NULL, ...){
    #default ice vehicle build for general use model
    #(ice_embrs1)

    #a lot here needs to replicated below in embrs_bev
    #don't think we can generalise at the moment
    #because of euro.class

    if(is.null(veh.type) || is.null(veh.wt)){
      stop("[embrs] all vehicles need both veh.type and veh.wt, see help?",
           call.=FALSE)
    }
    if(is.null(eng.fuel) || is.null(euro.class)){
      stop("[embrs] all ice vehicles...() need both eng.fuel and euro.class, see help?",
           call.=FALSE)
    }
    if(!tolower(method) %in% c("beddows", "embrs1", "embrs2")){
      stop("[embrs] unknown vehicle emission model, see help?",
           call.=FALSE)
    }

    .veh <- list(
      args = list(
        n = n,
        veh.type = veh.type,
        veh.wt = veh.wt,
        eng.type = "ice",
        eng.fuel = eng.fuel,
        euro.class = euro.class,
        brk.regen = FALSE
      ),
      #embrs1 method
      funs = list(
        ef.exh.pm2.5 = ef_vein_eea_exhaust_pm2.5,
        ef.brake.pm2.5 = ef_embrs1_brake_pm2.5,
        ef.tyre.pm2.5 = ef_embrs1_tyre_pm2.5,
        ef.road.pm2.5 = ef_embrs1_road_pm2.5,
        ef.resusp.pm2.5 = ef_embrs1_resusp_pm2.5,
        ef.exh.pm10 = ef_vein_eea_exhaust_pm10,
        ef.brake.pm10 = ef_embrs1_brake_pm10,
        ef.tyre.pm10 = ef_embrs1_tyre_pm10,
        ef.road.pm10 = ef_embrs1_road_pm10,
        ef.resusp.pm10 = ef_embrs1_resusp_pm10
      )
    )
    if(tolower(method)=="beddows"){
      .veh$funs <- list(
        ef.exh.pm2.5 = ef_vein_eea_exhaust_pm2.5,
        ef.brake.pm2.5 = ef_beddows_brake_pm2.5,
        ef.tyre.pm2.5 = ef_beddows_tyre_pm2.5,
        ef.road.pm2.5 = ef_beddows_road_pm2.5,
        ef.resusp.pm2.5 = ef_beddows_resusp_pm2.5,
        ef.exh.pm10 = ef_vein_eea_exhaust_pm10,
        ef.brake.pm10 = ef_beddows_brake_pm10,
        ef.tyre.pm10 = ef_beddows_tyre_pm10,
        ef.road.pm10 = ef_beddows_road_pm10,
        ef.resusp.pm10 = ef_beddows_resusp_pm10
      )
    }
    if(tolower(method)=="embrs2"){
      .veh$funs <- list(
        ef.exh.pm2.5 = ef_vein_eea_exhaust_pm2.5,
        ef.brake.pm2.5 = ef_embrs2_brake_pm2.5,
        ef.tyre.pm2.5 = ef_embrs2_tyre_pm2.5,
        ef.road.pm2.5 = ef_embrs1_road_pm2.5,
        ef.resusp.pm2.5 = ef_embrs1_resusp_pm2.5,
        ef.exh.pm10 = ef_vein_eea_exhaust_pm10,
        ef.brake.pm10 = ef_embrs2_brake_pm10,
        ef.tyre.pm10 = ef_embrs2_tyre_pm10,
        ef.road.pm10 = ef_embrs1_road_pm10,
        ef.resusp.pm10 = ef_embrs1_resusp_pm10
      )
    }

    ######################
    #failed!
    # trying to reduce the number of
    # warnings when missing args trigger
    # a warning
    #############################
    #last.warning <- last.warning[!duplicated(last.warning)]
    embrs_vehicle(name=name, x=.veh, ...)
  }


#splatted function
#' @rdname embrs.local
#' @export

embrs_bev <-
  function(veh.wt = NULL, veh.type= NULL,
           n = 1, method = "embrs1", name = NULL, ...){
    #default electric vehicle build
    #ice without the exhaust emissions

    if(is.null(veh.type) || is.null(veh.wt)){
      stop("[embrs] all vehicles need both veh.type and veh.wt, see help?",
           call.=FALSE)
    }
    if(!tolower(method) %in% c("beddows", "embrs1", "embrs2")){
      stop("[embrs] unknown vehicle emission model, see help?",
           call.=FALSE)
    }
    ############################
    #to think about
    ############################
    #error if fuel not electric?
    #euro.class might be trickier?

    .veh <- list(
      args = list(
        n = n,
        veh.type = "bus",
        veh.wt = veh.wt,
        eng.type = "electric",
        eng.fuel = "electric",
        brk.regen = FALSE
      ),
      #embrs1 method
      funs = list(
        ef.brake.pm2.5 = ef_embrs1_brake_pm2.5,
        ef.tyre.pm2.5 = ef_embrs1_tyre_pm2.5,
        ef.road.pm2.5 = ef_embrs1_road_pm2.5,
        ef.resusp.pm2.5 = ef_embrs1_resusp_pm2.5,
        ef.brake.pm10 = ef_embrs1_brake_pm10,
        ef.tyre.pm10 = ef_embrs1_tyre_pm10,
        ef.road.pm10 = ef_embrs1_road_pm10,
        ef.resusp.pm10 = ef_embrs1_resusp_pm10
      )
    )
    if(tolower(method)=="beddows"){
      .veh$funs <- list(
        ef.brake.pm2.5 = ef_beddows_brake_pm2.5,
        ef.tyre.pm2.5 = ef_beddows_tyre_pm2.5,
        ef.road.pm2.5 = ef_beddows_road_pm2.5,
        ef.resusp.pm2.5 = ef_beddows_resusp_pm2.5,
        ef.brake.pm10 = ef_beddows_brake_pm10,
        ef.tyre.pm10 = ef_beddows_tyre_pm10,
        ef.road.pm10 = ef_beddows_road_pm10,
        ef.resusp.pm10 = ef_beddows_resusp_pm10
      )
    }
    if(tolower(method)=="embrs2"){
      .veh$funs <- list(
        ef.exh.pm2.5 = ef_vein_eea_exhaust_pm2.5,
        ef.brake.pm2.5 = ef_embrs2_brake_pm2.5,
        ef.tyre.pm2.5 = ef_embrs2_tyre_pm2.5,
        ef.road.pm2.5 = ef_embrs1_road_pm2.5,
        ef.resusp.pm2.5 = ef_embrs1_resusp_pm2.5,
        ef.exh.pm10 = ef_vein_eea_exhaust_pm10,
        ef.brake.pm10 = ef_embrs2_brake_pm10,
        ef.tyre.pm10 = ef_embrs2_tyre_pm10,
        ef.road.pm10 = ef_embrs1_road_pm10,
        ef.resusp.pm10 = ef_embrs1_resusp_pm10
      )
    }
    ######################
    #failed!
    # trying to reduce the number of
    # warnings when missing args trigger
    # a warning
    #############################
    #last.warning <- last.warning[!duplicated(last.warning)]
    embrs_vehicle(name=name, x=.veh, ...)
  }











###############################
#currently unexported functions
###############################

###############################
#embrs_vehicle
##############################

#workhorse function for making
#vehicle/fleet objects

####################
#to do
####################

#doc args

#to think about
######################
#embrs_vehicle to handle embrs vehicle fleets and vehicles
#embrs_vehicle(name="hi", bus_bev(veh.wt=10000)[[1]][[1]], ef.brake.pm10 = NULL)[[1]][[1]]$fun


embrs_vehicle <-
  function(name = NULL, x = NULL, ...){

    .obj <- list(...)
    #print(.obj)
    if(is.null(x)){
      stop("nothing to work with!")
    }
    #could work up to handle x = embrs object?
    ## print(names(x))
    ## x = embrs_fleet_vehicle[[1]][[1]]
    obj <- list(args=list(),
                funs=list()
    )
    #was thinking of adding docs as extra case
    #BUT ending up doing it via ef_...(verbose=TRUE)
    test <- grep("^ef[.]", tolower(names(.obj)))
    .ef <- if(length(test)>0) {
      .obj[test]
    } else {
      list()
    }
    test <- which(!1:length(.obj) %in% test)
    .obj <- .obj[test]
    #x args and funs to output
    if(is.list(x$args)){
      obj$args <- modifyList(obj$args, x$args)
    }
    if(is.list(x$funs)){
      obj$funs <- modifyList(obj$funs, x$funs)
    }
    #... args and .ef to output
    if(is.list(.obj)){
      obj$args <- modifyList(obj$args, .obj)
    }
    if(is.list(.ef)){
      obj$funs <- modifyList(obj$funs, .ef)
    }

    name <- if(is.null(name)){
      if(obj$args$eng.fuel==obj$args$eng.type){
        paste(obj$args$eng.fuel, obj$args$veh.type, sep=".")
      } else {
        paste(obj$args$eng.fuel,  obj$args$eng.type, obj$args$veh.type, sep=".")
      }
    } else { as.character(name) }
    obj$args$name <- name

    out <- obj
    class(out) <- "embrs_vehicle"
    out <- list(out)
    names(out)[1] <- out[[1]]$args$name
    out <- list(fleet=out)
    class(out) <- c("embrs", "fleet")
    out
  }


#############################
#route_route
#############################

#used to build route_objects

####################
#to do
####################

#doc args

embrs_route <-
  function(name = NULL, route.def = NULL, route.source = NULL,
           route.dist = 1, route.slope = 0, ...){
    if(is.null(route.def) || is.null(route.source)){
      stop("[embrs] route_...() needs route.def and route.source [see help]",
           call. = FALSE)
    }
    obj <- list(args=list(name=name,
                          route.def=route.def,
                          route.source=route.source,
                          route.dist=route.dist,
                          route.slope=route.slope,
                          ...))
    if(is.null(obj$args$name)){
      obj$args$name <- paste(obj$args$route.source,  obj$args$route.def, sep=".")
    }
    class(obj) <- "embrs_route"
    out <- list(obj)
    names(out)[1] <- out[[1]]$args$name
    out <- list(routes=out)
    class(out) <- c("embrs", "routes")
    out
  }
